src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs

```diff
@@ -251,6 +271,135 @@ private void AnalyzeFormatArgument(OperationAnalysisContext context, IOperation
             }
         }
 
+        /// <summary>
+        /// Does the text have valid braces? (no unclosed braces, no braces without an opening, and no unescaped braces)
+        /// </summary>
+        /// <param name="formatter">The text to check.</param>
+        /// <returns>When true braces are valid, false otherwise.</returns>
+        private static bool IsValidBraces(LogValuesFormatter formatter)
+        {
```

Comment:
I think this parsing logic can be simplified without using place holders at all. Just by scanning the text one time. The code [https://source.dot.net/#Microsoft.Extensions.Logging.Abstractions/LogValuesFormatter.cs,88](https://source.dot.net/#Microsoft.Extensions.Logging.Abstractions/LogValuesFormatter.cs,88) can give some idea how we can do that. This code is not exact what we need but we can use similar idea. 

Let me know if you need any help writing it.

Reply:
Can you try/test the code:

```C#
        public static bool IsValidMessageTemplate(string messageTemplate)
        {
            if (messageTemplate is null)
            {
                return false;
            }

            if (messageTemplate.Length == 0)
            {
                return true;
            }

            int index = 0;
            bool leftBrace = false;

            while (index < messageTemplate.Length)
            {
                if (messageTemplate[index] == '{')
                {
                    if (index < messageTemplate.Length - 1 && messageTemplate[index + 1] == '{')
                    {
                        index++;
                    }
                    else if (leftBrace)
                    {
                        return false;
                    }
                    else
                    {
                        leftBrace = true;
                    }
                }
                else if (messageTemplate[index] == '}')
                {
                    if (leftBrace)
                    {
                        leftBrace = false;
                    }
                    else if (index < messageTemplate.Length - 1 && messageTemplate[index + 1] == '}')
                    {
                        index++;
                    }
                    else
                    {
                        return false;
                    }
                }

                index++;
            }

            return !leftBrace;
        }
```

Reply:
https://github.com/dotnet/roslyn-analyzers/pull/7286#discussion_r1715996888 basically a leetcode problem 😅